cmd/harmony/config_migrations.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/harmony-one/harmony/api/service/legacysync"
+	nodeconfig "github.com/harmony-one/harmony/internal/configs/node"
+	goversion "github.com/hashicorp/go-version"
+	"github.com/pelletier/go-toml"
+)
+
+const legacyConfigVersion = "1.0.4"
+
+func migrateConf(confBytes []byte) (harmonyConfig, string, error) {
+	var (
+		migratedFrom string
+	)
+	confTree, err := toml.LoadBytes(confBytes)
+	if err != nil {
+		return harmonyConfig{}, "", fmt.Errorf("config file parse error - %s", err.Error())
+	}
+	confVersion, found := confTree.Get("Version").(string)
+	if !found {
+		return harmonyConfig{}, "", errors.New("config file invalid - no version entry found")
+	}
+	migratedFrom = confVersion
+	if confVersion != tomlConfigVersion {
+		Ver, err := goversion.NewVersion(confVersion)
+		if err != nil {
+			return harmonyConfig{}, "", fmt.Errorf("invalid or missing config file version - '%s'", confVersion)
+		}
+		legacyVer, _ := goversion.NewVersion(legacyConfigVersion)
+		migrationKey := confVersion
+		if Ver.LessThan(legacyVer) {
+			migrationKey = legacyConfigVersion
+		}
+
+		migration, found := migrations[migrationKey]
+
+		// Version does not match any of the migration criteria
+		if !found {
+			return harmonyConfig{}, "", fmt.Errorf("unrecognized config version - %s", confVersion)
+		}
+
+		for confVersion != tomlConfigVersion {
+			confTree = migration(confTree)
+			confVersion = confTree.Get("Version").(string)
+			migration = migrations[confVersion]
+		}
+	}
+
+	// At this point we must be at current config version so
+	// we can safely unmarshal it
+	var config harmonyConfig
+	if err := confTree.Unmarshal(&config); err != nil {
+		return harmonyConfig{}, "", err
+	}
+	return config, migratedFrom, nil
+}
+
+var (
+	migrations = make(map[string]configMigrationFunc)
+)
+
+type configMigrationFunc func(*toml.Tree) *toml.Tree
+
+func init() {
+	migrations["1.0.4"] = func(confTree *toml.Tree) *toml.Tree {
+		ntStr := confTree.Get("Network.NetworkType").(string)
+		nt := parseNetworkType(ntStr)
+
+		defDNSSyncConf := getDefaultDNSSyncConfig(nt)
+
+		zoneField := confTree.Get("Network.DNSZone")
+		if zone, ok := zoneField.(string); ok {
+			confTree.Set("DNSSync.Zone", zone)
+		}
+
+		var port = int64(nodeconfig.DefaultDNSPort)
+		portField := confTree.Get("Network.DNSPort")
+		if p, ok := portField.(int64); ok {
+			if p != nodeconfig.DefaultLegacyDNSPort {
```

Comment:
Yes, you are right, will fix it and add a test for this case. Just to be absolutely clear - we subtract 3000 from legacy configs with the exception of the port being == DefaultDNSPort which is 6000 at the moment ?

Reply:
It is not expected that any node will run legacy port as 6000. Even if they config the port as 6000, it will be treated as 3000. So it's just fine to subtract 3000 for all cases.

Reply:
> * For each test case, have config file and cli command containing flags as test data.
> * Running CLI command with the config file will give expected node running behavior. (E.g., run with dns.serverPort will have node listen on specified server port. You can simply print that message in a customized branch.)
> * All test cases shall be reproducible.
> 
> I did some test cases long time ago, probably you can take a look:
> 
> * https://gist.github.com/JackyWYX/2554d561ccca74617d68a74a53ca64e3
> * https://gist.github.com/JackyWYX/adae08dd1984a2943c10a5358aa4b59a
> 

Thank you, that was helpful. I think I have a general Idea now what is required. I will conduct the tests today and post the result